File: CONADEA/app_conadea/lib/data/services/curso_service.dart
import 'dart:convert';

import '../models/curso.dart';
import 'api_client.dart';
import 'auth_service.dart';

/// Cursos reales (Backend/src/Controllers/CursoController.php). Crear es
/// exclusivo del rol Administrador; listar/obtener funciona para cualquier
/// rol autenticado — el token JWT ya viaja con el rol en ambos casos.
class CursoService {
  CursoService({ApiClient? apiClient, AuthService? authService})
      : _api = apiClient ?? ApiClient(),
        _authService = authService ?? AuthService();

  final ApiClient _api;
  final AuthService _authService;

  /// [datos] es todo lo del formulario menos los archivos (icono, título,
  /// descripción, lecciones, quiz); [imagenPath] es la ruta local del
  /// archivo elegido con la cámara o la galería, y [videosLecciones] trae
  /// (opcionalmente) la ruta local del video de cada lección, indexada por
  /// su posición en `datos['lecciones']` — el video es opcional por
  /// lección, así que no todos los índices tienen que estar presentes.
  /// Viaja todo en una sola petición multipart: el Backend guarda la
  /// imagen y los videos en Backend/uploads/cursos/{id}/... una vez que
  /// sabe el id del curso (y de cada lección).
  Future<void> crearCurso({
    required Map<String, dynamic> datos,
    required String imagenPath,
    Map<int, String> videosLecciones = const {},
  }) async {
    final token = await _authService.obtenerToken();
    await _api.postMultipart(
      '/cursos',
      fields: {'data': jsonEncode(datos)},
      files: {
        'imagen': imagenPath,
        for (final entry in videosLecciones.entries) 'leccion_video_${entry.key}': entry.value,
      },
      token: token,
    );
  }

  /// Catálogo completo (sin lecciones/quiz — GET /cursos solo manda lo
  /// general, para no cargar de más una lista).
  Future<List<Curso>> listarCursos() async {
    final token = await _authService.obtenerToken();
    final response = await _api.get('/cursos', token: token);
    final data = response['data'] as List<dynamic>;
    return data.map((e) => Curso.fromJson(e as Map<String, dynamic>)).toList();
  }

  /// Curso completo, con lecciones y quiz — para abrir el detalle.
  Future<Curso> obtenerCurso(int id) async {
    final token = await _authService.obtenerToken();
    final response = await _api.get('/cursos/$id', token: token);
    return Curso.fromJson(response['data'] as Map<String, dynamic>);
  }
}

File: CONADEA/Frontend/src/services/cursoService.js
import api from './api';

// Backend/src/Controllers/CursoController.php — crear() es exclusivo del
// rol Administrador (ver Authorize(['Administrador']) en el backend).
export default {
    crear(datos, imagenFile, videosLecciones = {}) {
        const formData = new FormData();
        formData.append('data', JSON.stringify(datos));
        formData.append('imagen', imagenFile);
        Object.entries(videosLecciones).forEach(([indice, archivo]) => {
            formData.append(`leccion_video_${indice}`, archivo);
        });

        // Deja que el navegador ponga el Content-Type con el boundary del
        // multipart — el header por defecto de la instancia (application/json)
        // rompería el envío de archivos si no se anula aquí.
        return api.post('/cursos', formData, {
            headers: { 'Content-Type': undefined }
        });
    },
    listar() {
        return api.get('/cursos');
    },
    obtener(id) {
        return api.get(`/cursos/${id}`);
    }
};
